### ISSUE TITLE
CUE and WebAssembly (Wasm) 

### ISSUE THREAD

**Issue by myitcv** [cue-lang/cue#2035]
This is an umbrella tracking issue for WebAssembly support in CUE. 

Details to follow, but initial exploration is linked to #2007.

**Comment by 4ad** [comment]
If you'd like to provide use cases and/or feedback on potential use of Wasm with CUE, please post to the discussion at https://github.com/cue-lang/cue/discussions/2045.

**Comment by sdboyer** [comment]
We have a solid use case for WASM with Thema - https://github.com/grafana/thema/issues/36.

Basically, we want a Thema-specific CUE playground-ish environment, which lets us present the user with all the basic thema operations. (Essentially, the operations exposed via the `thema` CLI).

Certainly, it would make [docs like this one](https://github.com/grafana/thema/blob/d936c1921950c4517cb9da1cb994340aa9cf6990/docs/go-quickstart.md) far more powerful - instead of asking people to run code generation, try out validation, etc., on their local, they can see it directly in the browser. But it would also enable us to do things like - just one example - update tutorials we write for configuring systems on the fly/in the browser, if their config file is written in Thema.

It's important to note that we need to have some Go logic in there, not just pure CUE, because we can't express all the constraints/invariants required by Thema in native CUE, and because our code generators are written in Go.

**Comment by myitcv** [comment]
For folks following along here, apologies we forgot to add notice of the [CUE + Web Assembly Focus Session Call](https://github.com/cue-lang/cue/discussions/2463). It was well advertised elsewhere, but we failed to join the dots here.

**Comment by 4ad** [comment]
As we discussed in several community calls, we're pausing Wasm developement until the Wasm ecosystem matures, in particular, we're waiting for the [Component Model](https://component-model.bytecodealliance.org).

**Comment by AndrewSav** [comment]
Will there be an alternate way to do custom functions?

**Comment by 4ad** [comment]
We are open to the idea, although we don't have a concrete design in mind yet. Care to share an experience report, or a simple feature request?

**Comment by yoktobit** [comment]
Would using an abstraction layer like extism help for integration of wasm? Then cue using wasm/wasi/wasipx wouldn't depend too much on their versions but more on the implementation in extism. https://github.com/extism/go-sdk

**Comment by keeganmccallum** [comment]
In case useful: https://github.com/dclareio/cue-wasm

**Comment by myitcv** [comment]
Quick heads-up for folks finding this thread via other routes. Per our [Community Call updates](https://github.com/cue-lang/cue/discussions/categories/community-calls), the CUE + WASM effort is on pause. In particular, the work to use WASM as a cross-language bridge for embedding the evaluator is currently on hold. 

Some context on that decision:

 • Ongoing flux in the WASM/WASI/WIT/component model standards landscape.  
 • The approach depends on [cue-lang/cue\#2007](https://github.com/cue-lang/cue/issues/2007), which we consider a critical requirement.  
 • That strand was also gated on eval v3 landing (which it now has).

As ever, we’re happy to re-evaluate priorities—if there’s renewed interest or concrete use cases here, please shout.

**Comment by mvdan** [comment]
@rogpeppe has just published a proposal for user-provided functions, which is very relevant here: https://github.com/cue-lang/cue/discussions/4293

Please leave any feedback in that thread, for the sake of keeping it all in one place.
